tokio-io/src/split.rs

```diff
@@ -0,0 +1,175 @@
+//! Split a single value implementing `AsyncRead + AsyncWrite` into separate
+//! `AsyncRead` and `AsyncWrite` handles.
+//!
+//! To restore this read/write object from its `split::ReadHalf` and
+//! `split::WriteHalf` use `unsplit`.
+
+use crate::{AsyncRead, AsyncWrite};
+
+use bytes::{Buf, BufMut};
+use futures_core::ready;
+use std::cell::UnsafeCell;
+use std::fmt;
+use std::io;
+use std::pin::Pin;
+use std::sync::atomic::AtomicU8;
+use std::sync::atomic::Ordering::{Acquire, Release};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+/// The readable half of a value returned from `split`.
+pub struct ReadHalf<T> {
+    inner: Arc<Inner<T>>,
+}
+
+/// The writable half of a value returned from `split`.
+pub struct WriteHalf<T> {
+    inner: Arc<Inner<T>>,
+}
+
+struct Inner<T> {
+    state: AtomicU8,
```

Comment:
Out of curiosity: any reason to use an `AtomicU8` rather than `AtomicBool`?

Reply:
no... i guess i probably should use bool. either way, this is hopefully a stopgap implementation.

Reply:
yeah, i thought `bool` conveyed intent more clearly, so i had assumed there was a reason for `u8` that i was missing :P

Reply:
bool makes sense tbh